src/esm_parser/provenance.py

```diff
@@ -183,14 +196,62 @@ def __setitem__(self, key, val):
         # standard ``__setitem__`` method from ``dict`` because that preserves all
         # object's methods and properties
         if isinstance(val, DictWithProvenance):
-            return super().__setitem__(key, val)
+            return super().__setitem__(key, val, *args, **kwargs)
         # If ``val`` is a ``dict`` but not a ``DictWithProvenance`` set the ``val``
         # as an instance of ``DictWithProvenance`` setting the ``provenance`` as
         # ``None``
-        elif isinstance(val, dict):
-            return super().__setitem__(key, DictWithProvenance(val, provenance=None))
+        #elif isinstance(val, dict):
+        #    return super().__setitem__(key, DictWithProvenance(val, provenance=None))
         # If the ``val`` is not a ``dict`` (it's a leaf of the dictionary tree) defines
         # ``self.provenance[key]`` as ``None``
         else:
-            self.provenance[key] = None
-            return super().__setitem__(key, val)
+            val_id = id(val)
+            self.provenance[key] = DictWithProvenance.leaf_id_provenance.get(val_id, None)
+            return super().__setitem__(key, val, *args, **kwargs)
+
+
+    def __delitem__(self, key, *args, **kwargs):
+
+        if isinstance(self, DictWithProvenance) and not isinstance(super().__getitem__(key), dict):
+            del self.provenance[key]
+
+        super().__delitem__(key, *args, **kwargs)
+
+
+    def update(self, dictionary, *args, **kwargs):
+        super().update(dictionary, *args, **kwargs)
+
+        for key, val in dictionary.items():
+            val_id = id(val)
+            if isinstance(dictionary, DictWithProvenance) and not isinstance(val, dict):
+                self.provenance[key] = dictionary.provenance.get(key,
+                    DictWithProvenance.leaf_id_provenance.get(val_id, None)
+                )
+            elif isinstance(dictionary, dict) and not isinstance(val, dict):
+                self.provenance[key] = DictWithProvenance.leaf_id_provenance.get(val_id, None)
+
+
+    def set_leaf_id_provenance(self, key):
+        # If it's a leaf
+        if not isinstance(super().__getitem__(key), DictWithProvenance):
+            val_id = id(super().__getitem__(key))
+            # Stores the provenance in a class variable, under an id key
+            DictWithProvenance.leaf_id_provenance[val_id] = self.provenance.get(key, None)
+
+
+def keep_id_provenance(func):
+    """
+    This is a decorator for recursive functions in esm_parser to preserve
+    provenanve. Better docstrings will come later...
+    """
```

Comment:
```suggestion
    """
    Decorator for recursive functions in esm_parser to preserve
    provenance.
    
    Recursive run functions in ``esm_parser`` are generally called on the innermost
    leaf. Here, we still run the function, but additionally store the output of the
    function into the `leaf_id_provenance` container so that provenance can be added
    to the result of the function call.
    
    Parameters
    ----------
    func : Callable
        The function to decorate
    """
```